src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -19,126 +19,65 @@
         {
-            Debug.Assert(process != null && stream != null && encoding != null && callback != null, "Invalid arguments!");
+            Debug.Assert(stream != null && encoding != null && callback != null, "Invalid arguments!");
             Debug.Assert(stream.CanRead, "Stream must be readable!");
             Debug.Assert(bufferSize > 0, "Invalid buffer size!");
 
-            _process = process;
             _stream = stream;
             _userCallBack = callback;
-            _encoding = encoding;
             _decoder = encoding.GetDecoder();
 
             if (bufferSize < MinBufferSize) bufferSize = MinBufferSize;
             _byteBuffer = new byte[bufferSize];
-            _maxCharsPerBuffer = encoding.GetMaxCharCount(bufferSize);
-            _charBuffer = new char[_maxCharsPerBuffer];
-
-            _eofEvent = new ManualResetEvent(false);
-            _messageQueue = new Queue<string>();
-        }
-
-        public void Close()
-        {
-            Dispose(true);
-        }
 
-        void IDisposable.Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
+            // This is the maximum number of chars we can get from one iteration in loop inside ReadBuffer.
+            // Used so ReadBuffer can tell when to copy data into a user's char[] directly, instead of our internal char[].
+            int maxCharsPerBuffer = encoding.GetMaxCharCount(bufferSize);
+            _charBuffer = new char[maxCharsPerBuffer];
 
-        private void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                if (_stream != null)
-                    _stream.Dispose();
-            }
-            if (_stream != null)
-            {
-                _stream = null;
-                _encoding = null;
-                _decoder = null;
-                _byteBuffer = null;
-                _charBuffer = null;
-            }
-
-            if (_eofEvent != null)
-            {
-                _eofEvent.Dispose();
-                _eofEvent = null;
-            }
-        }
-
-        public Encoding CurrentEncoding
-        {
-            get { return _encoding; }
-        }
-
-        public Stream BaseStream
-        {
-            get { return _stream; }
+            _cts = new CancellationTokenSource();
+            _messageQueue = new Queue<string>();
         }
 
         // User calls BeginRead to start the asynchronous read
         internal void BeginReadLine()
         {
-            if (_cancelOperation)
-            {
-                _cancelOperation = false;
-            }
-
             if (_sb == null)
             {
                 _sb = new StringBuilder(DefaultBufferSize);
-                _stream.ReadAsync(_byteBuffer, 0, _byteBuffer.Length).ContinueWith(ReadBuffer, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
+                _readToBufferTask = Task.Run(new Action(ReadBuffer), _cts.Token);
```

Comment:
I don't think we want to pass the token to the task like this.  Doing so could cause the task to end in the Canceled state, which would then cause the Wait on it in WaitUntilEOF to throw an exception, which I believe would then propagate out of WaitForExit.  Better to just not make the task cancelable.